datacompy/spark/sql.py

```diff
@@ -0,0 +1,1210 @@
+    ----------
+    filename : str
+        The file that contains the template.  Will automagically prepend the
+        templates directory before opening
+    fields : list
+        Fields to be rendered out in the template
+
+    Returns
+    -------
+    str
+        The fully rendered out file.
+    """
+    this_dir = os.path.dirname(os.path.realpath(__file__))
+    with open(os.path.join(this_dir, "..", "templates", filename)) as file_open:
+        return file_open.read().format(*fields)
+
+
+def columns_equal(
+    dataframe: "pyspark.sql.DataFrame",
+    col_1: str,
+    col_2: str,
+    col_match: str,
+    rel_tol: float = 0,
+    abs_tol: float = 0,
+    ignore_spaces: bool = False,
+    ignore_case: bool = False,
+) -> "pyspark.sql.DataFrame":
+    """Compares two columns from a dataframe, returning a True/False series,
+    with the same index as column 1.
+
+    - Two nulls (np.nan) will evaluate to True.
+    - A null and a non-null value will evaluate to False.
+    - Numeric values will use the relative and absolute tolerances.
+    - Decimal values (decimal.Decimal) will attempt to be converted to floats
+      before comparing
+    - Non-numeric values (i.e. where np.isclose can't be used) will just
+      trigger True on two nulls or exact matches.
+
+    Parameters
+    ----------
+    dataframe: pyspark.sql.DataFrame
+        DataFrame to do comparison on
+    col_1 : str
+        The first column to look at
+    col_2 : str
+        The second column
+    col_match : str
+        The matching column denoting if the compare was a match or not
+    rel_tol : float, optional
+        Relative tolerance
+    abs_tol : float, optional
+        Absolute tolerance
+    ignore_spaces : bool, optional
+        Flag to strip whitespace (including newlines) from string columns
+    ignore_case : bool, optional
+        Flag to ignore the case of string columns
+
+    Returns
+    -------
+    pyspark.sql.DataFrame
+        A column of boolean values are added.  True == the values match, False == the
+        values don't match.
+    """
+    base_dtype, compare_dtype = _get_column_dtypes(dataframe, col_1, col_2)
+    if _is_comparable(base_dtype, compare_dtype):
+        if (base_dtype in NUMERIC_SPARK_TYPES) and (
+            compare_dtype in NUMERIC_SPARK_TYPES
+        ):  # numeric tolerance comparison
+            dataframe = dataframe.withColumn(
+                col_match,
+                when(
+                    (col(col_1).eqNullSafe(col(col_2)))
+                    | (
+                        abs(col(col_1) - col(col_2))
+                        <= lit(abs_tol) + (lit(rel_tol) * abs(col(col_2)))
+                    ),
+                    # corner case of col1 != NaN and col2 == Nan returns True incorrectly
+                    when(
+                        (isnan(col(col_1)) == False)  # noqa: E712
+                        & (isnan(col(col_2)) == True),  # noqa: E712
+                        lit(False),
```

Comment:
@rhaffar I'm trying to remember (tackled this a while back) but it might have something to do with some how nans are handled. Take following dataframe:

```bash
+---+---+
|  x|  y|
+---+---+
|NaN|NaN|
|NaN|1.0|
|1.0|1.0|
|1.0|NaN|
+---+---+
```

If you run the following: 
```python
col1 = "x"
col2 = "y"
rel_tol = 0
abs_tol = 0
sdf.select(
    col_1,
    col_2,
    (abs(col(col_1) - col(col_2))).alias("1"),
    (lit(abs_tol) + (lit(rel_tol) * abs(col(col_2)))).alias("2"),
    (abs(col(col_1) - col(col_2)) <= lit(abs_tol) + (lit(rel_tol) * abs(col(col_2)))).alias("bool_check")
)
``` 
you end up with: 
```
+---+---+---+---+----------+
|  x|  y|  1|  2|bool_check|
+---+---+---+---+----------+
|NaN|NaN|NaN|NaN|      true|
|NaN|1.0|NaN|0.0|     false|
|1.0|1.0|0.0|0.0|      true|
|1.0|NaN|NaN|NaN|      true|
+---+---+---+---+----------+
```

If I'm remembering right, it is the 4th case above which says `1 == NaN` which isn't right. Its because the calculation ends up yield `NaN <= NaN` which is true but it's a very specific case. Hope I explained this OK, its just very weird.

Reply:
Ahhhh I see now - that is a weird issue indeed - I thought any comparison involving a `NaN` would automatically be false. Good to know though - thanks Faisal